panels/user-accounts/um-app-permissions.c

```diff
@@ -0,0 +1,795 @@
+/* um-app-permissions.c
+ *
+ * Copyright 2018 Endless, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include <libeos-parental-controls/app-filter.h>
+#include <flatpak.h>
+#include <gio/gio.h>
+#include <gio/gdesktopappinfo.h>
+#include <strings.h>
+
+#include "gs-content-rating.h"
+
+#include "um-app-permissions.h"
+
+struct _UmAppPermissions
+{
+  GtkGrid     parent_instance;
+
+  GMenu      *age_menu;
+  GtkSwitch  *allow_installation_switch;
+  GtkListBox *listbox;
+  GtkButton  *restriction_button;
+  GtkPopover *restriction_popover;
+
+  FlatpakInstallation *system_installation;
+  FlatpakInstallation *user_installation;
+
+  GSimpleActionGroup *action_group;
+
+  ActUser    *user;
+
+  GHashTable *blacklisted_apps;
+  GListStore *apps;
+
+  GCancellable *cancellable;
+  EpcAppFilter *filter;
+  guint         selected_age;
+
+  CcShell *shell;
+
+  guint          blacklist_apps_source_id;
+};
+
+static gboolean blacklist_apps_cb (gpointer data);
+
+static gint compare_app_info_cb (gconstpointer a,
+                                 gconstpointer b,
+                                 gpointer      user_data);
+
+static void on_allow_installation_switch_active_changed_cb (GtkSwitch        *s,
+                                                            GParamSpec       *pspec,
+                                                            UmAppPermissions *self);
+
+static void on_set_age_action_activated (GSimpleAction *action,
+                                         GVariant      *param,
+                                         gpointer       user_data);
+
+G_DEFINE_TYPE (UmAppPermissions, um_app_permissions, GTK_TYPE_GRID)
+
+enum
+{
+  PROP_0,
+  PROP_USER,
+  N_PROPS
+};
+
+static GParamSpec *properties [N_PROPS];
+
+static const GActionEntry actions[] = {
+  { "set-age", on_set_age_action_activated, "u" }
+};
+
+static const gchar * const oars_categories[] =
```

Comment:
While I obviously agree this should happen, it would require some refactoring in `gs-content-rating.c` since it doesn't even share it with its internal structures. And I'd rather avoid touching the copy-pasted code.

Reply:
I suggest leaving a `FIXME` comment here or filing a follow-up issue then, so this doesn’t get forgotten.

Reply:
I’ve pushed a fixup which adds a fixme.